### PR TITLE
[build] [coqide] Provide correct platform-specific hooks.

### LINKED ISSUE #14800: CoqIDE in v8.14 / master does not run on Windows
I built the v8.14 branch on native Windows using coq platform (tweaking 3 files in the platform).  When I run the resulting CoqIDE, I see the following. `Reset` immediately brings up the same dialog.  Debug output at bottom.  FWIW, note the CoqIDE version `8.14+alpha` in the Messages panel.

If I build v8.13.2 similarly, CoqIDE comes up just fine.

Perhaps someone else could try building and running this?  (At the moment, I can't run CoqIDE on Windows with `dune exec -- dev/shim/coqide-prelude` either.  More about that later.)

image

```
$ coqide -debug
(Spawn  ,616) EXEC: C:\cygwin64_coq2\home\jfehr\.opam\_coq-platform_.2021.03.0\bin\coqidetop.opt.exe -main-channel 127.0.0.1:52753:52754 -control-channel 127.0.0.1:52751:52752 --xml_format=Ppcmds -async-proofs off -debug
[DEBUG] Start eval_call Init None
[DEBUG] End eval_call
[DEBUG] End of Coqide.main
[DEBUG] Signal ignored (normal if Win32)
[DEBUG] Coqtop reader failed, resetting: Coq.TubeError("ERR")
Gtk-Message: GtkDialog mapped without a transient parent. This is discouraged.
[pid 204] <-- Init None
[pid 204] --> GOOD 1
```

### PR DESCRIPTION
**Pull request by ejgallego**
Actually, I disabled this to see if these hooks were used, but forgot
to enable them back.

Fixes #14800 .

Could be related to #14865 #14866 #14867 .

cc: @JasonGross @jfehrle

### CODE REVIEW

**Comment by SkySkimmer** [comment]
Are you supposed to have appveyor checks on your commits?

**Comment by ejgallego** [comment]
> Are you supposed to have appveyor checks on your commits?

I'm including #12425 so an artifact is created. By the way that PR is getting ready if someone would like to assign it.

**Comment by jfehrle** [comment]
@ejgallego The windows job failed.

**Comment by ejgallego** [comment]
> @ejgallego The windows job failed.

Indeed, the first build we did was not actually selecting the windows code (brain fade I had), the second rightly failed as we didn't include the low-level windows primitives in the build.

Let's see what happens with this one, if that words should be (finally) a good candidate to test.

File: ide/coqide/dune
Comment (gares) [inline ide/coqide/dune:32]: this file includes windows.h, it can only be used on windows

**Comment by ejgallego** [comment]
@jfehrle , it would be interesting if you could pinpoint which of the hacks in `coqide_WIN32.ml.in` is the one fixing #14800 

@JasonGross , can you check this PR against #14865 , #14866 #14867 ?

File: ide/coqide/dune
Comment (gares) [inline ide/coqide/dune:51]: instead of copying the files by hand, couldn't the ocaml program just spit the gtk backend name and then use that value and the standard `copy` dune directive?

Comment (SkySkimmer) [reply ide/coqide/dune:51]: If we use the os instead of graphics  name, there should be a dune variable for it

Comment (ejgallego) [reply ide/coqide/dune:51]: Dune variables are derived from `ocamlc -config`, which doesn't cover this use case well; see the relevant upstream bugs such as https://github.com/ocaml/ocaml/issues/10613

Comment (ejgallego) [reply ide/coqide/dune:51]: > instead of copying the files by hand, couldn't the ocaml program just spit the gtk backend name and then use that value and the standard `copy` dune directive?

I don't know.

**Comment by jfehrle** [comment]
Side question: what does `flock` do in the dune output (when VERBOSE=1)?  Outside of this PR, I've had problems where files in the source tree become very hard to remove/change.  If it means to lock a file in the file system, that might explain those problems.  It probably also means that dune doesn't always unlock the files, e.g. perhaps when the dune script is terminated with control-C.  My gut feel is that file locking is not a great (or common) design choice for this reason (hard to ensure that locks are released under all error conditions).

```
flock .dune.lock dune build --display quiet --profile=dev _build/default/plugins/funind/funind_plugin.cmxs
touch _build/default/plugins/btauto/btauto_plugin.cmxs
flock .dune.lock dune build --display quiet --profile=dev _build/default/plugins/nsatz/nsatz_plugin.cmxs
touch _build/default/plugins/funind/funind_plugin.cmxs
```

**Comment by SkySkimmer** [comment]
flock ensures that only 1 dune process is active at a time even if you call make with -j (because parallel dune calls is not supported).

> Outside of this PR, I've had problems where files in the source tree become very hard to remove/change.

Certainly unrelated.

**Comment by jfehrle** [comment]
> it would be interesting if you could pinpoint which of the hacks in `coqide_WIN32.ml.in` is the one fixing #14800

For a socket-related issue, the obvious choice is this line.  I verified that removing it brings back the problem.

`  Coq.gio_channel_of_descr_socket := Glib.Io.channel_of_descr_socket;`

However, dune is missing the dependency for this file--I had to do a `rm -rf *`, etc. to get an updated build.  Would you add that dependency (and any similar missing dependencies) in #14873?  (Perhaps it's possible to extract all the declared dependencies in Dune and compare that to the list of files present in git?)

**Comment by ejgallego** [comment]
Can we use zulip for these kind of unrelated discussions ?

**Comment by ejgallego** [comment]
Afaict the dependency on the file is declared correctly

**Comment by jfehrle** [comment]
> Can we use zulip for these kind of unrelated discussions ?

I don't use zulip regularly.  Given the time zone difference, it's no faster than email and it seems more ephemeral.

**Comment by ejgallego** [comment]
> I don't use zulip regularly. Given the time zone difference, it's no faster than email and it seems more ephemeral.

It is not OK to hijack PRs with unrelated discussions.

**Comment by ejgallego** [comment]
Actually we were aware of this problem, see https://github.com/coq/coq/issues/9885#issuecomment-490535777 and https://github.com/coq/coq/issues/9885#issuecomment-491091963 ; so I was puzzled coqide was not working on master as we had made it work.

Unfortunately it seems we forgot to open the issue, and the build remained broken for another 30 months O_O

**Comment by jfehrle** [comment]
> Unfortunately it seems we forgot to open the issue, and the build remained broken for another 30 months O_O

I don't see how you come to this conclusion.  #9885 was tagged for 8.10.  The 8.13 release doesn't have the issue.  But it's not surprising that some details fall through the cracks now and then in complex software like Coq.

**Comment by ejgallego** [comment]
> The 8.13 release doesn't have the issue.

The Windows dune build for CoqIDE has been broken for 8.10-8.14 so far, that's easy to see O_O

**Comment by MSoegtropIMC** [comment]
> The Windows dune build for CoqIDE has been broken for 8.10-8.14 so far, that's easy to see O_O

Yes, but up to 8.13 inclusive, dune was not the default build system.

File: ide/coqide/gen_ide_specific.ml
Comment (SkySkimmer) [inline ide/coqide/gen_ide_specific.ml:26]:
```suggestion
let () = main ()
```

**Review by SkySkimmer — APPROVED** [review]

File: ide/coqide/dune
Comment (SkySkimmer) [inline ide/coqide/dune:37]: Changed to use this copy method (allowing to use `copy#` for the ml file)
Testing says that dune understands the dependencies for `(action (copy#))` even though `%{read}` isn't allowed if we try to make it explicit with a `(deps)`.

**Comment by SkySkimmer** [comment]
@coq/build-maintainers this needs an assignee who isn't me since I modified the PR.

**Comment by ppedrot** [comment]
If this is only a matter of bureaucracy, I can assign since this is (also) a CoqIDE PR.

**Comment by SkySkimmer** [comment]
@ppedrot ok for me

**Review by ppedrot — APPROVED** [review]

**Comment by ppedrot** [comment]
Waiting for input from @JasonGross before merging.

File: ide/coqide/dune
Comment (ejgallego) [reply ide/coqide/dune:37]: Umm, in principle that should work but I dunno if we are walking a bit of an untested path here.

Dune 3.0 will support `%{read}` and `:include` everywhere [including deps], but for the 2.0 series, there are still some restrictions regarding dynamic build rules so I dunno; I guess we will see if this works or gives some trouble.

Comment (ejgallego) [reply ide/coqide/dune:37]: In fact, in 3.0 we could just have configure generate the stanza itself, and use include.

Comment (ejgallego) [reply ide/coqide/dune:37]: Another downside of this methods [which is not a big deal but should be pointed out] is that splitting coqide to its own package becomes harder; as we now have to take care of the config file.

Using ml libraries to manage this stuff allows us to reuse more infrastructure.

Comment (ejgallego) [reply ide/coqide/dune:37]: Yeah, coqide packages now have to run configure, I don't like that.

**Comment by ejgallego** [comment]
> Yes, but up to 8.13 inclusive, dune was not the default build system.

That doesn't mean the bug was there for all this time.

File: ide/coqide/dune
Comment (ejgallego) [reply ide/coqide/dune:37]: After thinking about this, please move the file generation from `./configure` to a small binary in `ide/coqide`.

**Review by ejgallego — COMMENTED** [review]
Note my change request: "After thinking about this, please move the file generation from ./configure to a small binary in ide/coqide."

**Comment by ejgallego** [comment]
I did the change myself; @ppedrot I suggest we merge this as it is a quite important blocker.

File: ide/coqide/gen_gtk_platform.ml
Comment (SkySkimmer) [inline ide/coqide/gen_gtk_platform.ml:7]: I feel like it's nicer to output to stdout, then only the dune file cares which filename is created

Comment (ejgallego) [reply ide/coqide/gen_gtk_platform.ml]: Indeed, I guess I'm too tired now, thanks.

Comment (ejgallego) [reply ide/coqide/gen_gtk_platform.ml]: I hope we can move that code completely out of Coq's configure by the way, #14189 almost does it.

**Comment by ppedrot** [comment]
@SkySkimmer Is that ready to merge then?

**Comment by SkySkimmer** [comment]
yes

**Pull request merged**
